File: src/CCVTAC.Console/PostProcessing/Deleter.cs
using System.IO;

namespace CCVTAC.Console.PostProcessing;

internal static class Deleter
{
    internal static void Run(
        IReadOnlyCollection<string> taggingSetFileNames,
        CollectionMetadata? collectionMetadata,
        string workingDirectory,
        Printer printer)
    {
        ImmutableList<string> collectionFileNames;
        var getFileResult = GetCollectionFiles(collectionMetadata, workingDirectory);
        if (getFileResult.IsSuccess)
        {
            var files = getFileResult.Value;
            collectionFileNames = files;

            printer.Debug($"Found {files.Count} collection files.");
        }
        else
        {
            collectionFileNames = [];
            printer.Warning(getFileResult.Errors.First().Message);
        }

        var allFileNames = taggingSetFileNames.Concat(collectionFileNames).ToImmutableList();

        if (allFileNames.IsEmpty)
        {
            printer.Warning("No files to delete were found.");
            return;
        }

        printer.Debug($"Deleting {allFileNames.Count} temporary files...");

        DeleteAll(allFileNames, printer);

        printer.Info("Deleted temporary files.");
    }

    internal static Result<ImmutableList<string>> GetCollectionFiles(
        CollectionMetadata? collectionMetadata,
        string workingDirectory)
    {
        if (collectionMetadata is null)
            return Result.Ok(ImmutableList<string>.Empty);

        try
        {
            var id = collectionMetadata.Value.Id;
            return Directory.GetFiles(workingDirectory, $"*{id}*").ToImmutableList();
        }
        catch (Exception ex)
        {
            return Result.Fail($"Error collecting filenames: {ex.Message}");
        }
    }

    private static void DeleteAll(IEnumerable<string> fileNames, Printer printer)
    {
        foreach (var fileName in fileNames)
        {
            try
            {
                File.Delete(fileName);

                printer.Debug($"• Deleted \"{fileName}\"");
            }
            catch (Exception ex)
            {
                printer.Error($"• Deletion error: {ex.Message}");
            }
        }
    }
}

File: src/CCVTAC.FSharp/PostProcessing/Deleter.fs
namespace CCVTAC.Console.PostProcessing

open CCVTAC.Console
open System.IO

module Deleter =
    let private getCollectionFiles
        (collectionMetadata: CollectionMetadata option)
        (workingDirectory: string)
        : Result<string array, string> =

        match collectionMetadata with
        | None -> Ok [||]
        | Some metadata ->
            try Ok (Directory.GetFiles(workingDirectory, $"*{metadata.Id}*"))
            with ex -> Error $"Error collecting filenames: {ex.Message}"

    let private deleteAll
        (fileNames: string array)
        (printer: Printer)
        : unit =

        fileNames
        |> Array.iter (fun fileName ->
            try
                File.Delete fileName
                printer.Debug($"• Deleted \"{fileName}\"")
            with
            | ex -> printer.Error($"• Deletion error: {ex.Message}")
        )

    let run
        (taggingSetFileNames: string seq)
        (collectionMetadata: CollectionMetadata option)
        (workingDirectory: string)
        (printer: Printer)
        : unit =

        let collectionFileNames =
            match getCollectionFiles collectionMetadata workingDirectory with
            | Ok files ->
                printer.Debug $"Found {files.Length} collection files."
                files
            | Error err ->
                printer.Warning err
                [||]

        let allFileNames = Seq.concat [taggingSetFileNames; collectionFileNames] |> Seq.toArray

        if Array.isEmpty allFileNames then
            printer.Warning "No files to delete were found."
        else
            printer.Debug $"Deleting {allFileNames.Length} temporary files..."
            deleteAll allFileNames printer
            printer.Info "Deleted temporary files."
